kevm-pyk/src/kevm_pyk/__main__.py

```diff
@@ -365,26 +362,50 @@ def _init_and_run_proof(claim: KClaim) -> tuple[bool, list[str] | None]:
 
             return passed, failure_log
 
-    results: list[tuple[bool, list[str] | None]]
-    if workers > 1:
-        with ProcessPool(ncpus=workers) as process_pool:
-            results = process_pool.map(_init_and_run_proof, claims)
-    else:
-        results = []
-        for claim in claims:
-            results.append(_init_and_run_proof(claim))
+    _LOGGER.info(f'Extracting claims from file: {spec_file}')
 
+    all_claims = kevm.get_claims(
+        spec_file,
+        spec_module_name=spec_module,
+        include_dirs=[Path(i) for i in includes],
+        md_selector=md_selector,
+        claim_labels=None,
+        exclude_claim_labels=exclude_claim_labels,
+    )
+
+    if all_claims is None:
+        raise ValueError(f'No claims found in file: {spec_file}')
+
+    all_claims_by_label: dict[str, KClaim] = {c.label: c for c in all_claims}
+    graph = {c.label: [f'{spec_module_name}.{d}' for d in c.dependencies] for c in all_claims}
+    _LOGGER.info(f'graph: {graph}')
+    topological_sorter = graphlib.TopologicalSorter(graph)
+    topological_sorter.prepare()
+    with ProcessPool(ncpus=workers) as process_pool:
+        selected_results = []
+        selected_claims = []
+        while topological_sorter.is_active():
+            ready = topological_sorter.get_ready()
+            _LOGGER.info(f'ready: {ready}')
+            curr_claim_list = [all_claims_by_label[label] for label in ready]
+            results = process_pool.map(_init_and_run_proof, curr_claim_list)
+            _LOGGER.info(f'done: {ready}')
+            for label in ready:
+                topological_sorter.done(label)
+            selected_results.extend(results)
+            selected_claims.extend(curr_claim_list)
+    _LOGGER.info('finished.')
```

Comment:
It looks like we lost some of the new logic from before which checked if there are multiple proofs to run or not. If not, then we don't start a process pool (overhead).

Reply:
It also looks like there  are some added logging messages for debugging that can be removed.